Migrate from ``autoflake``, ``black``, ``isort``, ``pyupgrade``, ``flake8`` and ``pydocstyle``, to ``ruff``

Pierre-Sassoulas opened the pull request:
Follow-up to #11896, this seems to be blocked by https://github.com/astral-sh/ruff/issues/4670 unless we're ready to accept a little churn. There are other tools that we could remove if we activate more ruff message, but I figured doing one at a time would help with the reviews.

nicoddemus commented:
Thanks @Pierre-Sassoulas!

I don't mind the churn, however I would prefer to see ruff replacing the other hooks in the same PR (specially black and autoflake, as well as whitespace, eol, etc), to avoid having multiples rounds of PRs which change a large portion of the code each.

Before merging, we can group all the automatic changes into a single commit and register that commit in https://github.com/pytest-dev/pytest/blob/main/.git-blame-ignore-revs.

bluetech commented on pyproject.toml:
```diff
@@ -123,3 +123,16 @@ target-version = ['py38']
 ignore = "W009"
+
+[tool.ruff]
+src = ["pytest"]
```
I think this should be `"src"` if I understand https://docs.astral.sh/ruff/settings/#src correctly.

Pierre-Sassoulas replied:
For sure, my bad.

bluetech commented on pyproject.toml:
```diff
@@ -123,3 +123,16 @@ target-version = ['py38']
+known-local-folder = ["pytest", "_pytest"]
+known-third-party = ["test_source", "test_excinfo"]
+force-sort-within-sections = true
+case-sensitive = false
```
This is the default, would remove.

Pierre-Sassoulas replied:
aa5f90166  (Changing to true does nothing by the way.)

bluetech commented on pyproject.toml:
```diff
@@ -123,3 +123,16 @@ target-version = ['py38']
+[tool.ruff.lint.isort]
+force-single-line = true
+known-local-folder = ["pytest", "_pytest"]
+known-third-party = ["test_source", "test_excinfo"]
```
Curious why this is needed (I know it's copied from isort config but maybe it's no longer needed?)

Pierre-Sassoulas replied:
Indeed, removing create no diff.

bluetech commented on pyproject.toml:
```diff
@@ -123,3 +123,16 @@ target-version = ['py38']
+force-single-line = true
+known-local-folder = ["pytest", "_pytest"]
+known-third-party = ["test_source", "test_excinfo"]
+force-sort-within-sections = true
```
reorder-imports separates `import`s and `from`s, so if want the same style we should keep it false (the default) no?

Pierre-Sassoulas replied:
Basically made some blind try to go around https://github.com/astral-sh/ruff/issues/4670 but the comment were not specific enough to be actionable, let's remove.

bluetech commented on pyproject.toml:
```diff
@@ -123,3 +123,16 @@ target-version = ['py38']
+    "I",  # isort
+]
+
+[tool.ruff.lint.isort]
```
I think [lines-after-imports](https://docs.astral.sh/ruff/settings/#isort-lines-after-imports) = 2 would also be nice (in theory at least...)

Pierre-Sassoulas replied:
Let's try it see 16479067c,

Pierre-Sassoulas commented:
I'm not sure it's possible to remove pre-commit's eol / whitespace checks because they act on all files when ruff fmt or black will only act on python files. Also @nicoddemus do we remove black for sure ? There's also blacken-doc it could maybe be replace by something like https://github.com/astral-sh/ruff/pull/8854 option... But I'm not sure it's equivalent because docstring != python code area in rst/md. Thus it would introduce two kind of formatting black for docstring and code example, and ruff for the code itself.

bluetech commented:
The isort looks good to me!

> There's also blacken-doc it could maybe be replace by something like https://github.com/astral-sh/ruff/pull/8854 option... But I'm not sure it's equivalent because docstring != python code area in rst/md. Thus it would introduce two kind of formatting black for docstring and code example, and ruff for the code itself.

From https://docs.astral.sh/ruff/formatter/#docstring-formatting it's also supposed to handle rst and md.

nicoddemus commented on .pre-commit-config.yaml:
```diff
@@ -20,14 +20,11 @@ repos:
+  hooks:
+    - id: ruff
+      args: ["--fix"]
 -   repo: https://github.com/PyCQA/flake8
```
I think we can remove `flake8` too. 👍

nicoddemus commented:
> From https://docs.astral.sh/ruff/formatter/#docstring-formatting it's also supposed to handle rst and md.

That's awesome. Did this work? I didn't notice any docs changing (which might mean it worked or did not even run).

Pierre-Sassoulas commented:
I rebased and squashed the isort commits together since they are reviewed now. Migrated autoflake / flake8 core. Still missing black blacken-doc and flake8-typing-imports atm

graingert commented:
Is flake8-typing-imports needed? ruff has support for flake8-type-checking https://github.com/astral-sh/ruff/pull/2147/files

Pierre-Sassoulas commented:
Apparently it's not supported yet : https://github.com/astral-sh/ruff/issues/2302

Pierre-Sassoulas commented:
We're [disabling a lot of message from flake8-docstrings](https://github.com/pytest-dev/pytest/blob/4546d5445aaefe6a03957db028c263521dfb5c4b/tox.ini#L195-L202) too.

bluetech commented on src/_pytest/_io/pprint.py:
```diff
@@ -376,7 +376,7 @@ def _pprint_simplenamespace(
         level: int,
     ) -> None:
-        if type(object) is _types.SimpleNamespace:
+        if isinstance(object, _types.SimpleNamespace):
```
This should stay as is if I understand the comment below correctly.

Pierre-Sassoulas replied:
Thank you for the reviews ! [287c651](https://github.com/pytest-dev/pytest/pull/11901/commits/287c651b7428a97777d0be4691b9a360facf8bef)

bluetech replied:
I'm thinking let's disable this rule entirely, generally it's a good rule but pytest can do weird low-level things and we usually know what we're doing when we use `type(..) is ...`.

Pierre-Sassoulas replied:
All right, I'll remove the manual fix for type / is instance and rebase everything cleanly.

bluetech commented on src/_pytest/setuponly.py:
```diff
@@ -74,7 +74,7 @@ def _show_fixture_action(
     tw.write(" " * 2 * scope_indent)
     tw.write(
-        "{step} {scope} {fixture}".format(
+        "{step} {scope} {fixture}".format(  # noqa UP032 (Readability)
```
```suggestion
        "{step} {scope} {fixture}".format(  # noqa: UP032 (Readability)
```

Pierre-Sassoulas commented:
Added ruff-format for code and doc. It changed some doc that blacken doc missed. I can produce a diff because a lot of things changed. I still need to check how many checks are actually used in `flake8-docstrings` to possibly remove flake8 entirely instead of simply disabling the core message.

bluetech commented:
Looks like most of the diff is from changing the max line length from 88 to 120. I think we should stick with 88 since that's the established style.

Pierre-Sassoulas commented:
Removed flake8-docstring, because it's a wrapper around pydocstyle, and pydocstyle explicitly asks to use ruff in https://github.com/PyCQA/pydocstyle/pull/658. Also the configuration given in the ruff thread about isort ``force-alphabetical-sort-within-sections`` option seems to work, so I used it.

I'm going to check how to format for 88 chars and not have warning for line longer than that next, thank you the pointer @bluetech .

Should I separate the manual fixes in another merge request ?

Pierre-Sassoulas commented:
Most of the diff now seems to be related to the newly autofixed rules from pydocstyle.

bluetech commented:
For the docstrings, we should go with D212 and disable D213, that would fit our current style better.

bluetech commented:
> Should I separate the manual fixes in another merge request ?

I think it's OK to do them here. I'll do a full look over once you're ready.

Pierre-Sassoulas commented:
I need to either add one of the autofix from pydocstyle in the unfixable list so it doesn't break the doctests, or exclude the code example from ruff-format (example_scripts/doctests/main.py, maybe others)

bluetech commented:
Looks at the docstring changes. Let's set

```
[tool.ruff.lint.pydocstyle]
convention = "pep257"
```

this allows to remove the ignores for the conflicting rules D203 and D213 since the convention implies that.

Also I see many changes to one-sentence docstrings

```

bluetech reviewed (commented):
Great work @Pierre-Sassoulas, went through everything and it looks good to me.

One thing though, I tried it and it seems like ruff format *doesn't* support formatting code snippets in rst and md files yet. It does support formatting rst and md snippets in docstrings but that's not what we need. So until ruff implements this (https://github.com/astral-sh/ruff/issues/8237), we should keep blacken-docs.

Pierre-Sassoulas commented on testing/test_assertrewrite.py:
```diff
@@ -1868,16 +1867,13 @@ def test_simple():
             id="assert statement not on own line",
         ),
         pytest.param(
-            b"def x():\n"
```
Oups, I missed this, I only searched for ``" f"`` and ``" "``, after fixing it would be ``b"def x():\n    assert (\n        1 + 2  # comment\n    )\n",`` not extraordinary, but better I guess ?

Pierre-Sassoulas replied:
Another way would be to change to:
```python
b"""def x():
    assert (
        1 + 2  # comment
    )
"""
```
But I don't know if it works for bytes.
```

bluetech replied:
Multiline b works but I think it will be annoying with the indentation.

Maybe just put `# fmt: off`/`# fmt: on` around them?

Pierre-Sassoulas replied:
The scope of ``# fmt: off/# fmt: on`` might not be exactly the same in ruff vs black, I had to "un-nest" the  existing ``fmt: off``.

Pierre-Sassoulas replied:
[3d53bbe](https://github.com/pytest-dev/pytest/pull/11901/commits/3d53bbebec39f0c7839fff747f0e5ac6a4dd546b)

Pierre-Sassoulas commented:
I rebased all the autofix commits to create the "one commit we can put in https://github.com/pytest-dev/pytest/blob/main/.git-blame-ignore-revs". Let me know what to do with all the manual fixes, squash all together or clean up the history.

Pierre-Sassoulas commented:
All right, I can squash everything then upgrade the `.git-blame-ignore-revs` (it will be two commits that need to be merged, not squashde or rebased so the sha is accurate). Thank you for the reviews, there was a lot of changes (especially at first with bad configurations)

nicoddemus commented:
@Pierre-Sassoulas thanks for the work on this! I think this should be backported to ease other backports.
